backend/src/main/java/com/example/backend/data_access/direction/DirectionDAO.java

```diff
@@ -0,0 +1,77 @@
+package com.example.backend.data_access.direction;
+
+import com.example.backend.entity.Location;
+import com.example.backend.entity.RouteDirection;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.*;
+
+public class DirectionDAO implements DirectionDataAccessInterface {
+    private final String directionsCsvFilename = "backend/src/main/java/com/example/backend/data/directions.csv";
+
+    /** Returns a map of all directions that run on a route, given its route tag
+     *
+     * @return mapping of direction tag to RouteDirection for the route **/
+    public HashMap<String, RouteDirection> getDirectionsByRouteTag(String routeTag) {
+        HashMap<String, RouteDirection> routeDirections = new HashMap<>();
+
+        // Used BufferedReader for massively improved performance
+        try (BufferedReader reader = new BufferedReader(new FileReader(directionsCsvFilename))) {
+            String line;
+            // Skip first row (headers)
+            reader.readLine();
+
+            while ((line = reader.readLine()) != null) {
+                String[] row = line.split(",");
+                if (Objects.equals(row[2], routeTag)) {
```

Comment:
How come you're using Objects.equals() and not String.equals()? 

Not necessarily an issue but if String.equals() doesn't work, it might indicate an issue with the data in the csv file and could be worth investigating.

Reply:
Oh also, iirc, assuming row[2] is a String, you can do row[2].equals(routeTag) instead. Not saying you should change the code, but just lyk for future reference.

Reply:
will change that just for code consistency